plugins/bnr-dp/fu-bnr-dp-firmware.c

```diff
@@ -0,0 +1,595 @@
+	csum = fu_bnr_dp_firmware_checksum_finish(csum);
+
+	return csum;
+}
+
+static gboolean
+fu_bnr_dp_firmware_payload_parse(FuBnrDpFirmware *self,
+				 GInputStream *stream,
+				 gsize payload_offset,
+				 GError **error)
+{
+	g_autoptr(GInputStream) payload_stream = NULL;
+	gsize streamsz = 0;
+	guint16 xml_checksum = 0;
+	guint16 crc = G_MAXUINT16;
+
+	payload_stream = fu_partial_input_stream_new(stream, payload_offset, G_MAXSIZE, error);
+	if (payload_stream == NULL)
+		return FALSE;
+
+	if (!fu_input_stream_size(payload_stream, &streamsz, error))
+		return FALSE;
+	if (self->payload_length != streamsz) {
+		g_set_error(
+		    error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_INVALID_FILE,
+		    "unexpected firmware payload length (header specified: %" G_GUINT64_FORMAT
+		    ", actual: %" G_GSIZE_FORMAT ")",
+		    self->payload_length,
+		    streamsz);
+		return FALSE;
+	}
+	if (streamsz != FU_BNR_DP_FIRMWARE_SIZE) {
+		g_set_error(
+		    error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_INVALID_FILE,
+		    "unexpected firmware payload length (must be: %d, actual: %" G_GSIZE_FORMAT ")",
+		    FU_BNR_DP_FIRMWARE_SIZE,
+		    streamsz);
+		return FALSE;
+	}
+
+	/* the XML header has a simple sum checksum for the payload */
+	if (!fu_bnr_dp_firmware_stream_checksum(payload_stream, &xml_checksum, error))
+		return FALSE;
+	if (self->payload_checksum != xml_checksum) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "checksum mismatch in firmware payload (XML header specified: 0x%X, "
+			    "actual: 0x%X)",
+			    self->payload_checksum,
+			    xml_checksum);
+		return FALSE;
+	}
+
+	/* we can do a CRC16 check on this type of payload as well */
+	if (!fu_input_stream_compute_crc16(payload_stream, FU_CRC_KIND_B16_BNR, &crc, error))
+		return FALSE;
+	if (crc != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "CRC mismatch in firmware payload: 0x%X",
+			    crc);
+		return FALSE;
+	}
+
+	/* discard the XML header and keep only the payload */
+	if (!fu_firmware_set_stream(FU_FIRMWARE(self), payload_stream, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_bnr_dp_firmware_parse(FuFirmware *firmware,
+			 GInputStream *stream,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuBnrDpFirmware *self = FU_BNR_DP_FIRMWARE(firmware);
+	guint8 byte = 0;
+	guint8 header_separator[] = {'\0'};
+	gsize separator_idx = 0;
+	g_autoptr(GBytes) header = NULL;
+	g_autoptr(XbBuilderSource) builder_source = xb_builder_source_new();
+	g_autoptr(XbBuilder) builder = xb_builder_new();
+	g_autoptr(XbSilo) silo = NULL;
+
+	if (!fu_input_stream_read_u8(stream, 0, &byte, error))
+		return FALSE;
+	/* simplistic but should generally be good enough to identify if a header is present */
+	if (byte != '<') {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "firmware file is missing XML header");
+		return FALSE;
+	}
```

Comment:
i don't think you need this; xb_builder_source_load_bytes() should be safe to load junk